torch/testing/_internal/custom_op_db.py

```diff
@@ -83,9 +86,114 @@ def numpy_take_impl(x, ind, ind_inv, dim):
     ind = to_numpy(ind)
     return torch.tensor(np.take_along_axis(x, ind, dim), device=device)
 
-@numpy_take.impl_meta()
-def numpy_take_meta(x, ind, ind_inv, dim):
-    return x
+@numpy_take.impl_fake()
+def numpy_take_fake(x, ind, ind_inv, dim):
+    return torch.empty_like(x)
+
+@custom_op('(Tensor x) -> Tensor', ns='_torch_testing')
+def numpy_nonzero(x):
+    ...
+
+@numpy_nonzero.impl(['cpu', 'cuda'])
+def numpy_nonzero_impl(x):
+    x_np = to_numpy(x)
+    res = np.stack(np.nonzero(x_np), axis=1)
+    if res.shape[0] <= 1:
+        raise RuntimeError("not supported")
+    return torch.tensor(res, device=x.device)
+
+@numpy_nonzero.impl_fake()
+def numpy_nonzero_fake(x):
+    ctx = torch._custom_op.get_ctx()
+    i0 = ctx.new_data_dependent_symint()
+    shape = [x.dim(), i0]
+    result = x.new_empty(shape, dtype=torch.long)
+    return result
```

Comment:
Opening up the floor for some bikeshedding. Here's some questions to get started:

ctx.new_data_dependent_symint()
- ctx.new_symint()? ctx.new_size()? ctx.new_unbacked_symint()?

`numpy_nonzero.impl_fake`
- Should we call our "fake implementation" "abstract implementation" instead, and rename this to `impl_abstract`? "fake implementation" sounds a bit confusing to me because the implementation ends up being registered for both Meta tensors and FakeTensors, but another way to think about this is FakeTensor is a superset of Meta tensors.

Reply:
Re (1), absent any pressure, I would line up the name with ShapeEnv, so `create_unbacked_symint`

Re (2), I actually kind of think we should lie and call it `impl_meta` anyway. I think my opinion is a bit different if we're going to support mixed-device custom ops, but if you're registering these implementations for Meta, they can't actually rely on a fake tensor mode being active (and thus always have to use `new_empty` and similar to make sure they are generated on the correct device.)

Reply:
> Re (1), absent any pressure, I would line up the name with ShapeEnv, so create_unbacked_symint

sgtm

> Re (2), I actually kind of think we should lie and call it impl_meta anyway. I think my opinion is a bit different if we're going to support mixed-device custom ops, but if you're registering these implementations for Meta, they can't actually rely on a fake tensor mode being active (and thus always have to use new_empty and similar to make sure they are generated on the correct device.)

I'm not sure we should lie, it would be bad if the user started creating new meta tensors inside the function (e.g. `torch.empty(3, 3, device=meta)`) because as it is currently set up: inside the FakeTensor rule, since the FakeTensorMode is active, we would interpret them as a FakeTensor wrapping a meta tensor

Reply:
OK, well, then we shouldn't call it fake or meta lol.

Reply:
Why not fake? Because it's not always a FakeTensor rule? (e.g. when registered to meta?)

Reply:
> Re (2), I actually kind of think we should lie and call it impl_meta anyway. I think my opinion is a bit different if we're going to support mixed-device custom ops, but if you're registering these implementations for Meta, they can't actually rely on a fake tensor mode being active (and thus always have to use new_empty and similar to make sure they are generated on the correct device.)

So, one idea for this is that for the meta implementation, we can activate FakeTensor mode and send FakeTensors that wrap meta tensors into the implementation. The benefit of this is that we're no longer lying about this being a fake impl and that the user doesn't need to worry about a third thing (abstract impl vs meta impl vs fake impl).

Reply:
> So, one idea for this is that for the meta implementation, we can activate FakeTensor mode and send FakeTensors that wrap meta tensors into the implementation. The benefit of this is that we're no longer lying about this being a fake impl and that the user doesn't need to worry about a third thing (abstract impl vs meta impl vs fake impl).

This doesn't seem possible because the Meta dispatch key is after Python

Reply:
Updated to `impl_abstract`.